src/Microsoft.OData.Core/Json/JsonValueUtilsAsync.cs

```diff
@@ -0,0 +1,491 @@
+                        currentIndex += bufferLength;
+                    }
+                    else
+                    {
+                        inputString.CopyTo(currentIndex, buffer.Value, 0, subStrLength);
+                        bufferIndex = subStrLength;
+                        currentIndex += subStrLength;
+                    }
+                }
+
+                // start writing escaped strings
+                for (; currentIndex < inputStringLength; currentIndex++)
+                {
+                    bufferIndex = EscapeAndWriteCharToBuffer(writer, inputString[currentIndex], buffer.Value, bufferIndex, stringEscapeOption);
+                }
+
+                // write any remaining chars to the writer
+                if (bufferIndex > 0)
+                {
+                    await writer.WriteAsync(buffer.Value, 0, bufferIndex).ConfigureAwait(false);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Escapes and writes a character array to a writer.
+        /// </summary>
+        /// <param name="writer">The text writer to write the output to.</param>
+        /// <param name="inputArray">Character array to write.</param>
+        /// <param name="inputArrayOffset">How many characters to skip in the input array.</param>
+        /// <param name="inputArrayCount">How many characters to write from the input array.</param>
+        /// <param name="stringEscapeOption">The string escape option.</param>
+        /// <param name="bufferPool">Character buffer pool.</param>
+        internal static async Task WriteEscapedCharArrayAsync(this TextWriter writer, char[] inputArray, int inputArrayOffset, int inputArrayCount, ODataStringEscapeOption stringEscapeOption, ICharArrayPool bufferPool)
+        {
+            int bufferIndex = 0;
+            char[] buffer = null;
+
+            buffer = BufferUtils.InitializeBufferIfRequired(bufferPool, buffer);
+
+            for (; inputArrayOffset < inputArrayCount; inputArrayOffset++)
+            {
+                bufferIndex = EscapeAndWriteCharToBuffer(writer, inputArray[inputArrayOffset], buffer, bufferIndex, stringEscapeOption);
+            }
+
+            // write remaining bytes in buffer
+            if (bufferIndex > 0)
+            {
+                await writer.WriteAsync(buffer, 0, bufferIndex).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Write the string value with quotes.
+        /// </summary>
+        /// <param name="writer">The text writer to write the output to.</param>
+        /// <param name="text">String value to be written.</param>
+        private static async Task WriteQuotedAsync(this TextWriter writer, string text)
+        {
+            await writer.WriteAsync(JsonConstants.QuoteCharacter).ConfigureAwait(false);
+            await writer.WriteAsync(text).ConfigureAwait(false);
+            await writer.WriteAsync(JsonConstants.QuoteCharacter).ConfigureAwait(false);

```

Comment:
I wonder what the trade-off would be between making 3 asynchronous write calls versus doing the concatenation and then making a single async write call:
   await writer.WriteAsync(String.Concat(JsonConstants.QuoteCharacter, text, JsonConstants.QuoteCharacter)).ConfigureAwait(false);

This seems to suggest, at least if the call result in Response.WriteAsync, the concatenation would be faster.  This would be interesting to benchmark.

Reply:
@mikepizzo I did some benchmarking and observed the following results that make a strong case for the concatenation

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
Intel Core i7-8665U CPU 1.90GHz (Coffee Lake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=5.0.103
  [Host]     : .NET Core 3.1.12 (CoreCLR 4.700.21.6504, CoreFX 4.700.21.6905), X64 RyuJIT
  Job-WIKFZT : .NET Core 3.1.12 (CoreCLR 4.700.21.6504, CoreFX 4.700.21.6905), X64 RyuJIT

InvocationCount=1  UnrollFactor=1

// Run 1

|                  Method |     Mean |   Error |  StdDev |
|------------------------ |---------:|--------:|--------:|
| WriteAsyncWithoutConcat | 202.4 ms | 3.93 ms | 3.49 ms |
|    WriteAsyncWithConcat | 120.4 ms | 2.27 ms | 1.89 ms |

// Run 2
|                  Method |     Mean |   Error |  StdDev |
|------------------------ |---------:|--------:|--------:|
| WriteAsyncWithoutConcat | 254.1 ms | 4.82 ms | 3.76 ms |
|    WriteAsyncWithConcat | 120.4 ms | 2.36 ms | 2.09 ms |

// Run 3
|                  Method |     Mean |   Error |  StdDev |
|------------------------ |---------:|--------:|--------:|
| WriteAsyncWithoutConcat | 163.1 ms | 2.42 ms | 2.02 ms |
|    WriteAsyncWithConcat | 121.3 ms | 2.03 ms | 2.49 ms |

Reply:
Awesome -- thanks for the analysis & update.  This will be a good opportunity for improvement throughout the stack, especially in perf-sensitive areas like reading/writing values.

---
In reply to: [582601314](https://github.com/OData/odata.net/pull/2002#discussion_r582601314) [](ancestors = 582601314)